### ISSUE TITLE
Events timestamp doesn't contain milliseconds

### ISSUE THREAD

**Issue by awwitecki** [splunk/splunk-library-javalogging#143]
Since version 1.8.0 events received by Splunk have timestamps with milliseconds set to 000. Hence, order of frequent events is distorted. We haven't changed anything in our logging but splunk-library-javalogging version from 1.7.3 to 1.8.0.

Sample log on Splunk (confidential data are blurred but they don't contain any time-related information):
![Screenshot 2020-03-19 at 17 54 50](https://user-images.githubusercontent.com/3371155/77093411-34b9d580-6a0b-11ea-904c-91cde0197f28.png)

For copy-paste:
_time: 2020-03-19T17:53:16.000+01:00
time: 1584635570.598 (Epoch for 2020-03-19T17:32:50.598+01:00 - there's a delay in logs transfer but it actually also quite nicely shows that time is not derived from events data)

**Comment by snorwin** [comment]
In the last release, the timestamp was moved from the meta-information (header) level to the event itself. However, if Splunk has not been explicitly configured, it ignores the timesamp in the event. The EventHeaderSerializer (#134) would give you full control on the meta-information of an HEC request and therefore you can set your timestamp in the header directly.

We have been using this feature for quite some time to eliminate the delay of the log transfer in order  to create more accurate metrics or to feed Splunk with historical data.

**Comment by hanswurscht** [comment]
Hi, 

can this be documented somewhere?
A correctly filled _time-field ist essential for Splunk (and other software).

Thanks!
- Lorenz

**Comment by snorwin** [comment]
I'm not sure if it was intended or accidentally changed.

**Comment by memarshall63** [comment]
Is there any update on this issue.   I've been wrestling with it for a couple days now.  The messages I see have an epoch time in a "time" field at the highest level, but for some reason Splunk doesn't pick that off as a valid time stamp.  My props.conf has:

SHOULD_LINEMERGE = false
TIME_FORMAT = %s.%3N
TIME_PREFIX = "time":"

but Splunk still doesn't capture the full time (with milliseconds).  I presume it's using the Splunk system time.  

I'd question my props.conf -- but when I export this traffic raw and then put it back through Add Data -- it works fine.  So I'm guessing it's HEC related.

Puzzling to me.   I get Splunk, Java is not in my wheelhouse.  If someone could explain a workaround or approach to me -- that'd be peachy.

MM.
![Screen Shot 2020-09-30 at 9 52 48 AM](https://user-images.githubusercontent.com/33108108/94694322-b8317900-0302-11eb-8897-241d7cf2539d.png)

**Comment by awwitecki** [comment]
I could only suggest you using Splunk Universal Forwarder as a side app for your application. There you may specify desired timestamp field with ease in `props.conf` per sourcetype, e.g.
```
[any_sourcetype]
TIMESTAMP_FIELDS=timestamp
(...)
```
In our project we decided to go this way as this also sped our services up a lot.

**Comment by memarshall63** [comment]
You're suggesting not using the library at all?  Just dropping a UF and monitoring the log files coming out of log4j?  The only reason I haven't done this is that they're running in docker containers -- not that it's impossible - just haven't plumb those lines yet.

**Comment by awwitecki** [comment]
Exactly! We were also refraining from going for this due to the same reason (Docker + k8s). At the end of the day we came out with a satisfiable solution with side containers. Bear in mind however that there are some corner cases when it comes to binding containers lifecycles and it depends on you and your environment what approach should be taken to handle them properly.

**Comment by memarshall63** [comment]
Thanks for the input.  I appreciate you taking the time.   I'll looking into alternative docker approaches.

**Comment by macrergate** [comment]
> However, if Splunk has not been explicitly configured, it ignores the timesamp in the event. 
> 

@snorwin , don't you know how to configure Splunk to use the timestamp in the event? I can't find anything.

**Comment by snorwin** [comment]
@macrergate we still use the EventHeaderSerializer of the PR #134 and set the time from the event in the header of the HEC request and Splunk automatically takes that header as the timestamp `_time`.

However, there seem to be other ways to configure the extraction of the timestamp from the event, (https://docs.splunk.com/Documentation/Splunk/7.3.6/Data/HowSplunkextractstimestamps). I never tried this and I'm not sure if this is working at all using this log4j appender.

**Comment by memarshall63** [comment]
Thanks for the input.  I've tried to extract the epoch timestamp from the message, but for some reason Splunk doesn't pull it correctly.  I've used: 

SHOULD_LINEMERGE = false
TIME_FORMAT = %s.%3N
TIME_PREFIX = "time":"

There's some chatter about HEC and timestamps and HTTP headers on the community boards, but I haven't read all the way through.  At the moment, my app team is using 1.7.3 of the library, which doesn't seem to add the "time" element.  The dev added our own timestamp, and we're able to extract that in PROPS, so we have something working, but no 1.8 upgrade path.   If anyone gets a definitive answer to this, perhaps a document on their approach could be offered.

**Comment by macrergate** [comment]
I'm on the same page as @memarshall63 - we are downgraded to 1.7.3

@zenmoto , if we could also ask you about the change? Why attribute time was moved from metadata to the event data? It looks like it was changed in the commit: [use gson for json serialization] (https://github.com/splunk/splunk-library-javalogging/commit/4e6102e08a098662dc3714027ea98bae202c3404)
Before this commit in HttpEventCollectorSender.java: line 250 was:
`putIfPresent(event,` MetadataTimeTag, String.format(Locale.US, "%.3f", eventInfo.getTime()));`

serialization was moved to HecJsonSerializer but without the logic above.

**Comment by hanswurscht** [comment]
So since version 1.9 is released now, this still seems to be an issue?
What do we need to implement in our java code to send the timestamp with millisecond precision correctly?

**Comment by sierralex58** [comment]
> So since version 1.9 is released now, this still seems to be an issue?
> What do we need to implement in our java code to send the timestamp with millisecond precision correctly?

dear hans

if you use log4j2 you could try the following pattern:

date_time="%d{ISO8601_OFFSET_DATE_TIME_HHCMM}"
date_time: 2021-04-23T15:01:49,233+02:00

which in my case, displays the ISO-time of the message when it was sent to splunk (not the index-time) including ms.

**Comment by hanswurscht** [comment]
> date_time="%d{ISO8601_OFFSET_DATE_TIME_HHCMM}"
> date_time: 2021-04-23T15:01:49,233+02:00
> 
> which in my case, displays the ISO-time of the message when it was sent to splunk (not the index-time) including ms.

I think my question was not accurate enough. 


In our testing, the timestamp display in the message of the log itself is with millisecond precision. The time, read by splunk an put into the `_time`-field has no millisecond precision. This is described in the first screenshot of this issue.

Does the configuration above (`date_time`) change this behavior, so that splunk extracts the milliseconds correctly to `_time`?

Thanks!

**Comment by snorwin** [comment]
Hi @hanswurscht 
We use custom key/value log format (`key1="value1" key2="value2"`) from which we extract the time stamp and set it with milliseconds in the `serializeEventHeader`:  

```
public class JsonEventSerializer implements EventHeaderSerializer {

    @Override
    public Map<String, Object> serializeEventHeader(HttpEventCollectorEventInfo eventInfo, final Map<String, Object> metadata) {
        Map<String, String> tags = parseTags(eventInfo.getMessage());
        try {
            Instant instant = Instant.from(DateTimeFormatter.ISO_OFFSET_DATE_TIME.parse(tags.get("date_time").replace(',', '.')));
            metadata.put(MetadataTags.TIME, String.format("%d.%03d", instant.getEpochSecond(), instant.getNano() / 1000000));
        } catch (Exception ignored) { /* nop */ }

        return metadata;
    }

    private Map<String, String> parseTags(String message) {
        Map<String, String> tags = new HashMap<>();

        try {
            // Parse message to map: key1="value1" key2="value2"
            Splitter.on(Pattern.compile("\"\\s(?=\\w+=\")"))
                    .trimResults()
                    .withKeyValueSeparator(Splitter.on("=").limit(2).trimResults().trimResults(CharMatcher.is('\"')))
                    .split(message)
                    .forEach((k, v) -> {
                        // CamelCase to snake_case conversion
                        k = Splitter.on(Pattern.compile("(?=[A-Z])")).splitToList(k).stream()
                                .map(String::toLowerCase)
                                .collect(Collectors.joining("_"));

                        tags.put(k, v);
                    });
        } catch (Exception ex) {
            tags.put("message", message);
        }

        return tags;
    }
}
```

In the `log4j2.xml` we then set the reference to the `EventHeaderSerializer`:
```
eventHeaderSerializer="com.exmaple.JsonEventSerializer"
```

Not sure if that helps to solve your issue.

Cheers,
Norwin

**Comment by fantavlik** [comment]
This issue was addressed in #165 - please see version 1.9.0 to take this change.
